### PR TITLE
umqtt.simple: Do the DNS resolution every time you connect

### PR DESCRIPTION
**Pull request by jcea**
Do the DNS resolucion everytime you connect.

This is specially relevant for client "umqtt.robust" because you could not have network access when instantiate the object, so subsequent connections tries (after network is available) will fail because the "cached" address is wrong.

Resolving DNS in each connection is not a problem because reconnections should be sporadic and your DNS server (typically your local router) most probably cached the DNS reply anyway.

### CODE REVIEW

**Comment by pfalcon** [comment]
> Resolving DNS in each connection is not a problem

But what if your DNS server went down, while everything else works? Then with you change, you (everyone!) is hosed, whereas with the code like it is now, it will likely continue to work (host addresses change quite rarely in general).

Bottom line: I wish someone else +1 this.

**Comment by pfalcon** [comment]
And while we're waiting for it, could you please format commit message consistently with the existing commits (just do "git log" and you'll see how they should be). Thanks.

**Comment by jcea** [comment]
DNS server failure should be both critical and temporal (if your DNS is failing, you sure are going to notice soon). In any case, with my proposal "connect()" will recover gracefully when your DNS is up again.

If you care about DNS dying, current code is even worse, because that "failure" will be cached forever unless you drop your MQTT object and create a new instance.

PS: I will be careful with future commits messages. Thanks for pointing out.

**Comment by pfalcon** [comment]
Suppose that sounds good. But you didn't test your change ;-). (Hint: see what params you pass at the new location of getaddrinfo())

**Comment by dpgeorge** [comment]
To make it even more robust you could catch any exceptions from the getaddrinfo and not update the cached address if the lookup fails.

**Comment by timpur** [comment]
Hi, I submitted #155. 

This request solves my issue that i stated. But @dmascord had a slightly different solution but practically the same. Only difference is, to cache the address? Also as dpgeorge said you could use the old address if getting a new one fails? Something to think about? Either way I'm happy with these changes.

**Comment by MrSurly** [comment]
Found this PR because I was going to submit something similar.  DNS resolution can be done in `__init__` , and re-done as needed, but shouldn't prevent `__init__` from succeeding.

**Comment by puuu** [comment]
I am also interested in this patch.

IMHO doing DNS lookup during initialization is bad. If the class has a connect() method, I expect, that everything before calling this method can be done without network connection.

I have a typical setup, the ESP8266 starts up, read a sensor, publish the sensor data via mqtt and go in deep sleep for a few minutes. Short before running the connect() method, the program checks and waits for network connection with the hope that the ESP already has connected during initialization and sensor read out. This should reduce the awake time. Unfortunately, I did not expected that the umqtt already need network during initialization for DNS lookup. This was hard to debug, because this only happens at the final sensor location (bad network condition, long connection time), where I do not have a serial debug connection ...

So, to summarize: +1 

@jcea, you are willing to fix rework this patch as suggested by @pfalcon ?
> could you please format commit message consistently with the existing commits (just do "git log" and you'll see how they should be)

>  with it connect() accesses undefined local variables server, port (which then may be randomly picked from the outer context).

**Comment by pfalcon** [comment]
@puuu , In case, feel free to submit an alternative PR, I'll merge it.

**Comment by puuu** [comment]
> @puuu , In case, feel free to submit an alternative PR, I'll merge it.

@pfalcon please see #191 .

**Comment by pfalcon** [comment]
Closing in favor of #191.
